### PR TITLE
Introduce encoding limits [TE-301]

### PR DESCRIPTION
**Pull request by akoptelov**
- Introduced new bounded encoding primitives
- Bounded encoding primitives are used to guard possibly unlimited data structures in P2P messages
- Boundaries are taken mostly from Tezos code

For reference, see https://simplestakingcom.atlassian.net/l/c/rwb1LGxf.

### CODE REVIEW

File: tezos/encoding/src/encoding.rs [hunk 1]
```diff
@@ -242,8 +251,13 @@ pub enum Encoding {
     /// Is the collection of fields.
     /// prefixed its length in bytes (4 Bytes), encoded as the concatenation of all the element in binary
     Dynamic(Box<Encoding>),
+    /// Is the collection of fields.
+    /// prefixed its length in bytes (4 Bytes), encoded as the concatenation of all the element in binary
+    BoundedDynamic(usize, Box<Encoding>),
     /// Represents fixed size block in binary encoding.
     Sized(usize, Box<Encoding>),
+    ///
```
Comment (tizoc) [inline tezos/encoding/src/encoding.rs:259]: Looks like a comment is missing here.

Comment (akoptelov) [reply tezos/encoding/src/encoding.rs:259]: Thanks for catching this.

File: tezos/messages/src/p2p/encoding/advertise.rs [hunk 2]
```diff
@@ -35,5 +36,11 @@ impl AdvertiseMessage {
 
 cached_data!(AdvertiseMessage, body);
 has_encoding!(AdvertiseMessage, ADVERTISE_MESSAGE_ENCODING, {
-    Encoding::Obj(vec![Field::new("id", Encoding::list(Encoding::String))])
+    Encoding::Obj(vec![Field::new(
+        "id",
+        Encoding::blist(
```
Comment (bkontur) [inline tezos/messages/src/p2p/encoding/advertise.rs:41]: @akoptelov 
I would prefer full name `blist` -> `bounded_list`,
ocaml code is full of one-letter variables and it is pain to read it sometimes

File: tezos/encoding/src/binary_writer.rs [hunk 3]
```diff
@@ -367,6 +407,31 @@ fn encode_value(data: &mut Vec<u8>, value: &Value, encoding: &Encoding) -> Resul
...
+            let bytes_sz = encode_value(data, value, dynamic_encoding)?;
+            if bytes_sz > *max {
+                return Err(bound_error(encoding, *max, bytes_sz));
+            }
+
+            // capture slice of buffer where List length was stored
+            let mut bytes_sz_slice =
+                &mut data[data_len_before_write..data_len_after_size_placeholder];
+            // update size
+            bytes_sz_slice.write_u32::<BigEndian>(bytes_sz as u32)?;
```
Comment (tizoc) [inline tezos/encoding/src/binary_writer.rs:427]: Is there a better way to perform this cast `bytes_sz as u32` (from `usize` to `u32`) ?

File: tezos/messages/src/p2p/encoding/limits.rs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+pub const MAX_PASS: usize = 8;
+
+/// Advertise list, max 100 items
+pub const ADVERTISE_ID_LIST_MAX_LENGTH: usize = 100;
+
+/// CurrentBranch history max length
+pub const CURRENT_BRANCH_HISTORY_MAX_LENGTH: usize = 200;
+
+/// Block header max length
+pub const BLOCK_HEADER_MAX_LENGTH: usize = 8 * 1024 * 1024;
```
Comment (bkontur) [inline tezos/messages/src/p2p/encoding/limits.rs:22]: @akoptelov 
could you please, add to every constant where we can find it in ocaml code?
like we talked about max operations yesterday please

Comment (tizoc) [reply tezos/messages/src/p2p/encoding/limits.rs:22]: Yes, was about to make that exact comment. At least the file and variable name containing the definition.

File: tezos/messages/src/p2p/encoding/advertise.rs (hunk 2 above)
Comment (akoptelov) [reply tezos/messages/src/p2p/encoding/advertise.rs:41]: OK.

File: tezos/encoding/src/binary_writer.rs (hunk 3 above)
Comment (akoptelov) [reply tezos/encoding/src/binary_writer.rs]: Sure.

File: tezos/messages/src/p2p/encoding/limits.rs (hunk 4 above)
Comment (akoptelov) [reply tezos/messages/src/p2p/encoding/limits.rs]: Done.

File: tezos/encoding/src/encoding.rs [hunk 5]
```diff
@@ -294,6 +323,13 @@ impl Encoding {
         Encoding::Dynamic(Box::new(encoding))
     }
 
+    /// Utility function to construct [Encoding::Dynamic] without the need
+    /// to manually create new [Box].
+    #[inline]
+    pub fn bdynamic(max: usize, encoding: Encoding) -> Encoding {
```
Comment (bkontur) [inline tezos/encoding/src/encoding.rs:329]: @akoptelov 
bdynamic -> bounded_dynamic?
please, check also all other shortcuts

**Review by bkontur — APPROVED** [review]
@akoptelov 
I guess, ok now,

@tizoc
Bruno, what about you?

**Comment by tizoc** [comment]
@bkontur my comments have been resolved, looks ok to me 👍

**Comment by bkontur** [comment]
> 
> 
> @bkontur my comments have been resolved, looks ok to me +1

@akoptelov 
please squash commit and rebase to develop and then merge

**Review by tizoc — APPROVED** [review]

**Comment by akoptelov** [comment]
Thanks, guys!

**Pull request merged**